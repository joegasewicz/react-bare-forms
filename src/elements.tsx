import {default as React} from "react";

import {IValidators} from "./validators";
import {
    IForm,
    IRadioGroupParentContext,
    RadioGroupContext,
    TypeRadioGroupMetadata,
} from "./form";
import {SubmitButton as _SubmitButton,} from "./field_classes/_SubmitButton";
import {
    CheckBoxField as _CheckBoxField,
    FileField as _FileField,
    InputField,
    RadioField as _RadioField,
    SelectField as _SelectField,
    TextAreaField as _TextAreaField,
} from "./field_classes";

/** @internal */
export enum FIELD_NAMES {
    TEXT = "text",
    EMAIL = "email",
    PASSWORD = "password",
    TEXTAREA = "textArea",
    RADIO  = "radio",
    CHECKBOX = "checkbox",
    SELECT = "select",
    FILE = "file",
}

export interface IFieldBase {
    /** The name of the form element (this should match the state property that you want be updated by this form element) */
    name: string;
    /** If *bare* (see {@link IForm.bare}) is set to true then *labeltext* will be inserted within **label** tags. */
    labeltext?: string;
    /** If *bare* (see {@link Form.bare}) is set to true then *hint* will be inserted within **small** tags. */
    hint?: string;
    /** These are the {@link IValidators} that you can pass in the validate the form element. */
    validators?: IValidators;
    /** Add additional css */
    className?: string;
}

export interface IField<T> extends React.InputHTMLAttributes<T> {
    /** The name of the form element (this should match the state property that you want be updated by this form element) */
    name: string;
    /** If *bare* (see {@link IForm.bare}) is set to true then *labeltext* will be inserted within **label** tags. */
    labeltext?: string;
    /** If *bare* (see {@link Form.bare}) is set to true then *hint* will be inserted within **small** tags. */
    hint?: string;
    /** These are the {@link IValidators} that you can pass in the validate the form element. */
    validators?: IValidators;
    /** Add additional css */
    className?: string;
}

export interface ITextInputField extends IField<HTMLInputElement> {
    /** The state property that gets updated by this input field */
    value: any;
}

export interface IPasswordField extends IField<HTMLInputElement> {
    /** The state property that gets updated by this input field */
    value: any;
}

export interface IEmailField extends IField<HTMLInputElement> {
    /** The state property that gets updated by this input field */
    value: any;
}

export interface ICheckBoxField extends IField<HTMLInputElement> {
    checked: boolean;
}

export interface ITextAreaField extends IField<HTMLTextAreaElement> {
    rows?: number;
    /** The state property that gets updated by this input field */
    value: any;
}

export interface IRadioField extends IField<HTMLInputElement> {
    checked: boolean;
    disabled?: boolean;
}

export interface IFileField extends IField<HTMLInputElement> {
    ref: React.RefObject<HTMLFormElement>;
}

export type TypeSelectCssSizeName = |"sm"|"default"|"lg";

export interface ISelectField extends Omit<IField<HTMLSelectElement>, "size"> {
    value: any;
    options: Array<string|{[k: string]: any}>;
    size?: TypeSelectCssSizeName;
    objectkey?: string;
    objectvalue?: string;
}

/**
 *
 * @param props {@link ITextInputField}
 * @example
 * ```
 *  import {TextInputField} from "react-base-forms"
 *
 *  const state = { username: "" }
 *
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <TextInputField
 *    value={state.username}
 *    name="username"
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <TextInputField
 *    value={state.username}
 *    name="username"
 *    hint="Needs to be at least 50 characters long"
 *    labeltext="Username"
 *  />
 * ```
 * @constructor
 */
export const TextInputField = (props: ITextInputField) => {
  const textInput = new InputField<ITextInputField>(FIELD_NAMES.TEXT, props);
  return textInput.create();
};

/**
 * @param props {@link IEmailField}
 * @example
 * ```
 *  import {EmailField} from "react-base-forms"
 *
 *  const state = { email: "" }
 *
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <EmailField
 *    value={state.email}
 *    name="email"
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <EmailField
 *    value={state.email}
 *    name="email"
 *    hint="Needs to be at least 50 characters long"
 *    labeltext="Username"
 *  />
 * ```
 * @constructor
 */
export const EmailField = (props: IEmailField) => {
    const emailInput = new InputField<IEmailField>(FIELD_NAMES.EMAIL, props);
    return emailInput.create();
};

/**
 *  The `PasswordField` works the same as the `EmailField` & `TextInputField`'s.
 * @param props {@link IPasswordField}
 * @example
 * ```
 *  import {areFieldsEqual, isFieldEmpty, PasswordField} from "react-base-forms";
 *
 *  const state = { password: "", confirmPassword: "" };
 *
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <PasswordField
 *    value={state.password}
 *    name="username"
 *    validators={[isFieldEmpty(8)]}
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <PasswordField
 *    value={state.confirmPassword}
 *    name="password"
 *    hint="Needs to be at least 8 characters long"
 *    labeltext="Password"
 *  />
 *  ```
 * Also we can create two *PasswordField* components to confirm passwords are equal. Please see
 * {@link areFieldsEqual} for more info.
 * The first *PasswordField* has has a *name* prop of **password** & the second *PasswordField* a name
 * prop of *confirmPassword*. Then we can add a {@link areFieldsEqual} validator to the *PasswordField*
 * with the *confirmPassword* name props (also notice how {@link areFieldsEqual} takes the first *PasswordField*
 * name as an argument).
 *
 * @example
 * ```
 * <PasswordField
 *  name="password"
 *  // other props...
 *
 * />
 *
 * <PasswordField
 *  name="confirmPassword"
 *  // other props...
 *  validators={[areFieldsEqual("password")]}
 * />
 * ```
 *
 * @constructor
 */
export const PasswordField = (props: IPasswordField) => {
  const passwordInput = new InputField<IPasswordField>(FIELD_NAMES.PASSWORD, props);
  return passwordInput.create();
};



/**
 * The CheckBoxField requires a *checked* prop instead of a *value* prop. See
 * {@link ICheckBoxField}.
 * @example
 * ```
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 *  import {CheckBoxField} from "react-base-forms";
 *
 *  const state = { password: "", confirmPassword: "" };
 *
 * <CheckBoxField
 *   name="terms"
 *   checked={state.terms}
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 * <CheckBoxField
 *   name="terms"
 *   checked={state.terms}
 *   hint="Click to agree"
 *   labeltext="Agree to terms & conditions"
 * />
 * ```
 * @param props {@link ICheckBoxField}
 * @constructor
 */
export const CheckBoxField = (props: ICheckBoxField) => {
  const checkBox = new _CheckBoxField<ICheckBoxField>(FIELD_NAMES.CHECKBOX, props);
  return checkBox.create();
};

/**
 * The TextAreaField takes in an extra prop of *row* which is a number & declares
 * the number of rows displayed by the textarea element. The TextAreaField accepts
 * all the {@link IField} props.
 * @example
 * ```
 *  *  import {CheckBoxField} from "react-base-forms";
 *
 *  const state = { about: "" };
 * // A bare form example ... remember to set the {@link Form.bare} property to `true`
 * <TextAreaField
 *    value={state.about}
 *    name="about"
 *    validators={[isFieldEmpty(20)]}
 * />
 *
 * // Example with Bootstrap styling (Bootstrap styling comes as default)
 *
 * <TextAreaField
 *    name="about"
 *    value={state.about}
 *    hint="Must be at least 20 characters"
 *    labeltext="About you..."
 *    validators={[isFieldEmpty(20)]}
 * />
 * ```
 * @param props {@link ITextInputField}
 * @constructor
 */
export const TextAreaField = (props: ITextAreaField) => {
  const textArea = new _TextAreaField<ITextAreaField>(FIELD_NAMES.TEXTAREA, props);
  return textArea.create();
};

export interface IRadioGroupProps {
    name: string;
    children: any;
}

/**
 * @props {@link IRadioGroupProps}
 * @param props The `RadioGroup` component takes a single props of `name`, which
 * must be a unique to a form. See {@link RadioField}.
 * @constructor
 *
 * @example
 * ```
 *  import {CheckBoxField} from "react-base-forms";
 *
 *  const state = { male: true, female: false };
 *
 *  <RadioGroup name="group1">
 *    // place RadioFields components here...
 *  </RadioGroup>
 * ```
 *
 */
export function RadioGroup(props: IRadioGroupProps) {
    const contextValue: IRadioGroupParentContext = {parent: {name: props.name}, children: props.children};
    return <RadioGroupContext.Provider value={contextValue as TypeRadioGroupMetadata}>{props.children}</RadioGroupContext.Provider>;
}

/**
 *
 * @param props {@link IRadioField}
 * @constructor
 * `RadioField` inputs are designed to be used with the {@link RadioGroup} component.
 *  To use this component, add or nest it within a {@link RadioGroup} component as children.
 *  It's possible to also use validators with a RadioGroup, as shown below:
 * @example
 * ```
 *  import {isRadioChecked, RadioField, RadioGroup} from "react-base-forms";
 *
 *  const state = { male: true, female: false };
 *
 *  <RadioGroup name="group1">
 *    <RadioField
 *      name="male"
 *      checked={state.male}
 *      hint="Click to agree"
 *      labeltext="Agree to terms & conditions"
 *    />
 *
 *    <RadioField
 *      name="female"
 *      checked={state.female}
 *      hint="Click to agree"
 *      labeltext="Agree to terms & conditions"
 *      validators={[isRadioChecked()]}
 *    />
 *
 *  </RadioGroup>
 * ```
 */
export const RadioField = (props: IRadioField) => {
    const radio = new _RadioField(FIELD_NAMES.RADIO, props);
    return radio.create();
};

/**
 * A component to render a select field element.
 * @param props {@link ISelectField}
 * @constructor
 * @example
 * ```
 *  import {SelectField} from "react-base-forms";
 *
 *  const state = { fruitChoice: "" };
 *
 * <SelectField
 *   size="lg"
 *   value={state.fruitChoice}
 *   name="fruitChoice"
 *   options={["banana", "apple", "orange"]}
 *  />
 * ```
 *
 * You can also pass an array of objects but you must use both the
 * *objectKey* & *objectvalue* props. the `objectKey` will update your state
 * value & the `objectvalue` is what is displayed to the user as an option.
 * @example
 * ```
 * // This is your option data
 * let selectData = [
 *   {id: 1, name: "first"},
 *   {id: 2, name: "second"},
 * ];
 * // The state which will receive the update
 * let state = {
 *    select_data_id: undefined as any,
 * };
 *
 * <SelectField
 *   size="lg"
 *   value={state.select_data_id}
 *   name="fruitChoice"
 *   objectkey="id" // Value will update state.select_data_id e.g *1, 2...*
 *   objectvalue="name" // Value will be displayed in the select field e.g *first, second...*
 *   options={selectData}
 * />
 * ```
 */
export const SelectField = (props: ISelectField) => {
    const select = new _SelectField(FIELD_NAMES.SELECT, props);
    return select.create();
};

/**
 * The SubmitButton only requires a text string as children props (see below example).
 * The SubmitButton will be disabled until all form fields are validated.
 * @param props {@link ISubmitButtonProps}
 * @constructor
 * @example
 * ```
 * import {SubmitButton} from "react-base-forms";
 *
 * <SubmitButton>Submit</SubmitButton>
 * ```
 */
export const FileField = React.forwardRef((props: IFileField, ref: React.RefObject<HTMLFormElement>|any) => {
    let _props = {...props, ref };
    const file = new _FileField<IFileField>(FIELD_NAMES.FILE, _props);
    return file.create();
});


export const SubmitButton = new _SubmitButton().create();
