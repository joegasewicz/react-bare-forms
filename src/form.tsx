import {default as React, useEffect, useState} from "react";
import {updateRadioGroupStateFromPassedInContext, updateParentState, _FieldEmptyErrorMsg} from "./core";
import {IValidation} from "./validators";
import {
    AbstractMetadata,
    Metadata,
    MetadataFile,
    MetadataGroup,
} from "./field_classes";
import { _noContextError } from "./core/_errors";
import {updateCursorPosCallback, updateQueryResultsCallback} from "./core/_handlers";

/** @internal */
export type TypeFieldValueTypes = "value"|"checked"|"file";
/** @internal */
export interface IFieldValues {
    currentValue: any;
    readonly type: TypeFieldValueTypes;
}
/** @internal */
export interface IFieldValidation {
    readonly name: string;
    validation: Array<IValidation>;
    fieldValues: IFieldValues;
    isTouched: boolean;
}
/** @internal */
export interface IInputFieldMetadata extends IFieldValidation {}
/** @internal */
export interface IFileMetaData extends IFieldValidation {
    readonly refName: string;
}
/** @internal */
export interface IRadioGroupChildren extends IFieldValidation {}
/** @internal */
export interface ICheckBoxesMetadata extends IFieldValidation {}
/** @internal */
export type TypeInputMetadata = AbstractMetadata<IInputFieldMetadata>;
/** @internal */
export type TypeFileMetadata = AbstractMetadata<IFileMetaData>;
/** @internal */
export type TypeRadioGroupMetadata = AbstractMetadata<IRadioGroupChildren>;
/** @internal */
export type TypeCheckboxesMetadata = AbstractMetadata<ICheckBoxesMetadata>;
/** @internal */
export type TypeFormMetadata =
    | TypeInputMetadata
    | TypeRadioGroupMetadata
    | TypeCheckboxesMetadata
    | TypeFileMetadata;

export interface IMetadata {
    radioGroups: TypeRadioGroupMetadata;
    inputs: TypeInputMetadata;
    files: TypeInputMetadata;
    checkboxes: TypeCheckboxesMetadata;

}
/** @internal **/
export enum METADATA_NAMES {
    INPUTS = "inputs",
    RADIO_GROUPS = "radioGroups",
    FILES = "files",
    CHECKBOXES = "checkboxes",
}
/**
 * @interface **IForm** Exported Form interface available to the caller. Contains all the properties required by
 * the Form *RBF* Form's component.
 */
export interface IForm extends React.FormHTMLAttributes<HTMLFormElement> {
    /** The passed in state from the parentName component */
    state: any;
    /** If the parentName component is a class component then the context must contain the parentName's **this** keyword. */
    context?: any;
    /** If the form's state is not directly contained at parentName's state root object, then formKey needs to represent the key. */
    readonly formKey?: string;
    /** Default is false. If set to true then the form state will be displayed just below the form component in `code` tags. */
    readonly debug?: boolean;
    /** Default is True. If set to false, the  validators will not be called until the form is submitted. */
    readonly dynamic?: boolean;
    /** The passed through props to the Form component */
    children?: any;
    /** Default set to false. If set to true then it will now use bootstrap styling or any extra elements. */
    readonly bare?: boolean;
    /** Callback function wil be called on form submission if all validators pass */
    readonly callback?: Function;
}
/** @internal */
export interface ICursorPositionState {
    cursorPosition: number;
    fieldName: string;
}
/** @internal */
export interface IQueryState {
    queryResults: Array<any>;
    showResults: boolean;
}
/** @internal */
export type TypeCursorPositionState = {[key: string]: ICursorPositionState};
/** @internal */
export type TypeQueryState = {[key: string]: IQueryState};

export interface IFormContext {
    readonly bare?: boolean;
    readonly debug?: boolean;
    readonly dynamic?: boolean;
    readonly formKey?: string;
    cursorPositions: TypeCursorPositionState;
    queryState: TypeQueryState;
    metadata: IMetadata;
    state: any;
    updateCursorPositionState: (fieldName: string, cursorPosition: number) => void;
    updateParentState?: (e: React.ChangeEvent<any>, name: string, formKey?: string) => void;
    updateRadioGroupStateFromPassedInContext?: (e: React.ChangeEvent<any>, name: string, radioGroup: any, formKey?: string) => void;
    updateQueryResultsState: (fieldName: string, showResults: boolean) => void;
}
/** @internal */
const INPUTS_STATE: TypeInputMetadata = {} as any;
/** @internal */
const RADIO_GROUPS_STATE: TypeRadioGroupMetadata = {} as any;
/** @internal */
const FILES_STATE: TypeFileMetadata = {} as any;
/** @internal */
const CHECKBOXES_STATE: TypeCheckboxesMetadata = {} as any;
/** @internal **/
const CURSOR_POSITION_STATE: TypeCursorPositionState = {};
/** @internal **/
const QUERY_STATE: TypeQueryState = {};
/** @internal */
const providerContext: IFormContext = {
    bare: false,
    state: {},
    formKey: null as any,
    debug: false,
    dynamic: true,
    cursorPositions: {},
    queryState: {},
    updateCursorPositionState: {} as any,
    updateQueryResultsState: {} as any,
    metadata: {
        inputs: null as any,
        radioGroups: null as any,
        files: null as any,
        checkboxes: null as any,
    },
};
/** @internal */
export interface IRadioGroupParentContext {
    parent?: { name: string };
    children?: any;
}
/** @internal */
export const RadioGroupContext = React.createContext<TypeRadioGroupMetadata>(RADIO_GROUPS_STATE);
/** @internal */
export const FormContext = React.createContext<IFormContext>(providerContext);
/** @internal */
export const FormProvider = FormContext.Provider;
/**
 * @var Access the form state within a React Context Api Consumer scope
 * @example For example:
 * ```
 * <Form state={this.state} context={this}>
 *      <FormConsumer>
 *      {({state}) => {
 *          return <div>{state} is equal {myState}</div>;
 *      }}
 *    </FormConsumer>
 * </Form>
 * ```
 * */
export const FormConsumer = FormContext.Consumer;

/** @internal */
export const handleSubmit = (props: IForm) =>
    (e: React.ChangeEvent<any>) => {
        e.preventDefault();
        const { callback, } = props;
        if(typeof callback === "function" && callback()) {
            callback();
        } else {
            // TODO throw error
        }
    };

/**
 * The main Form component that is required to wrap all *RBF* components.
 * If the component that uses the Form component is a functional component then
 * only the state props & state update function returned from the useState hook are
 * required. If you are calling Form component from a class component then you must 
 * pass your local context or `this` keyword to
 * the `context` prop.
 * @param props {@link IForm}
 *
 * An example using *RBF* Form component in a functional component
 * @example
 * ```
 * // Minimal setup for a RBF's Form component
 *
 *  const myState = {
 *      username: '',
 *  }
 *  const [state, setState] = React.useState(state);
 *  <Form state={myState} context={setState}></Form>
 * ```
 *
 * To use *RBF* Form component from a class component you must pass in your
 * local state of `this` keyword.
 * @example
 * ```
 * // Minimal setup for a RBF's Form component for a class component
 *
 *  this.state = { // in the constructor
 *      username: '',
 *  }
 *
 *  <Form state={this.state} context={this}></Form>
 * ```
 * @constructor
 */
export const Form = (props: IForm) => {
    const [parentState, setParentState] = useState(props.state);
    const [context, _] = useState(providerContext);
    const [inputState, updateInputState] = useState(INPUTS_STATE);
    const [checkboxesState, updateCheckboxesState] = useState(CHECKBOXES_STATE);
    const [fileState, updateFileState] = useState(FILES_STATE);
    const [radioState, updateRadioState] = useState(RADIO_GROUPS_STATE);
    const [cursorPositionState, updateCursorPositionState] = useState<TypeCursorPositionState>(CURSOR_POSITION_STATE);
    const [queryState, updateQueryState] = useState<TypeQueryState>(QUERY_STATE);
    // If the parentName component is a class component, then the state needs to be updated from the parentName context
    // Functions component must use useState hook. See https://joegasewicz.github.io/react-bare-forms/modules/_form_.html
    if(props.context && "setState" in props.context) {
        useEffect(() => {
            props.context.setState({
                ...parentState,
            });
        }, [parentState]);
    } else if (props.context) {
        useEffect(() => {
            props.context({
                ...parentState,
            });
        }, [parentState]);
    } else {
        throw new _FieldEmptyErrorMsg(_noContextError);
    }
    // State Hooks
    const _providerContext: IFormContext = {
        bare: props.bare || context.bare,
        state: props.state,
        formKey: props.formKey,
        cursorPositions: cursorPositionState,
        queryState: queryState,
        debug: props.debug || context.debug,
        dynamic: props.dynamic || context.dynamic,
        updateParentState: updateParentState(parentState, setParentState, props.formKey),
        updateRadioGroupStateFromPassedInContext: updateRadioGroupStateFromPassedInContext(parentState, setParentState, props.formKey),
        updateCursorPositionState: updateCursorPosCallback(cursorPositionState!, updateCursorPositionState),
        updateQueryResultsState: updateQueryResultsCallback(queryState, updateQueryState),
        metadata: {
            [METADATA_NAMES.INPUTS]: new Metadata<IInputFieldMetadata>(
                inputState as {},
                updateInputState,
                METADATA_NAMES.INPUTS,
            ),
            [METADATA_NAMES.CHECKBOXES]: new Metadata<ICheckBoxesMetadata>(
                checkboxesState as {},
                updateCheckboxesState,
                METADATA_NAMES.CHECKBOXES,
            ),
            [METADATA_NAMES.FILES]: new MetadataFile<IFileMetaData>(
                fileState as {},
                updateFileState,
                METADATA_NAMES.FILES,
            ),
            [METADATA_NAMES.RADIO_GROUPS]: new MetadataGroup<IRadioGroupChildren>(
                radioState as any,
                updateRadioState,
                METADATA_NAMES.RADIO_GROUPS,
            ),
        },
    };
    return (
        <FormProvider value={_providerContext}>
            <form onSubmit={handleSubmit(props)}>{props.children}</form>
        </FormProvider>
    );
};
