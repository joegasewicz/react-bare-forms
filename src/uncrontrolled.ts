/////////////////////////////////////////////////////////
// Public uncontrolled React Component Helper functions
/////////////////////////////////////////////////////////
import {default as React} from "react";
import {IFile} from "./core";

/**
 * RBF's provides a function that returns a React ref to access your file object.
 * To use, simply assign the returned ref from the createFileRef function to a
 * variable & pass this variable to FileField's ref prop.
 * @returns
 * @example
 * ```
 *
 *  import {createFileRef, FileField, isFile, getFileFromRef} from "react-bare-forms";
 *
 *  const myFileRef = createFileRef();
 *
 *  <FileField
 *     ref={myFileRef}
 *     hint="Must be a file"
 *     labelText="Upload your file"
 *     name="myFileTest"
 *     validators={[isFile()]}
 *  />
 *
 * ```
 */
export function createFileRef(): React.RefObject<HTMLFormElement> {
    return React.createRef<HTMLFormElement>();
}

/**
 * To get a file object back from the react ref, you can use `getFileFromRef` function.
 * @params The React ref returned from {@link createFileRef}
 * @returns
 * @example
 * ```
 *
 *  import {createFileRef, FileField, isFile, getFileFromRef} from "react-bare-forms";
 *
 *  const myFileRef = createFileRef();
 *
 *  <FileField
 *     ref={myFileRef}
 *     hint="Must be a file"
 *     labelText="Upload your file"
 *     name="myFileTest"
 *     validators={[isFile()]}
 *  />
 *
 *  let fileData = {
 *      myFile: getFileFromRef(myFileRef)
 *  };
 *
 *  // fileData.myFile is your Javascript File object.
 *
 * ```
 */
export function getFileFromRef(fileRef: React.RefObject<HTMLFormElement>): IFile|null {
    if(fileRef && (fileRef as any).current && Object.keys((fileRef as any).current.files).length) {
        return (fileRef as any).current.files[0];
    } else {
        return null;
    }
}
