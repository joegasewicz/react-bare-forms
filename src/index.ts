import {Form, FormConsumer} from "./form";
import {
    CheckBoxField,
    EmailField,
    FileField,
    ICheckBoxField,
    IEmailField,
    IField,
    IPasswordField,
    IRadioField, ISelectField,
    ITextAreaField,
    ITextInputField,
    IDatePicker,
    IQueryField,
    PasswordField,
    RadioField,
    RadioGroup,
    SelectField,
    SubmitButton,
    TextAreaField,
    TextInputField,
    DatePickerField,
    QueryInputField,
} from "./elements";
import {
    areFieldsEqual,
    customValidator,
    isChecked,
    isEmailValid,
    isFieldEmpty,
    isFile,
    isRadioChecked,
    isValidDate,
} from "./validators";
import {createFileRef, getFileFromRef} from "./uncrontrolled";


export {
    // Uncontrolled
    createFileRef,
    getFileFromRef,
    // Validators
    areFieldsEqual,
    customValidator,
    isChecked,
    isEmailValid,
    isFieldEmpty,
    isFile,
    isRadioChecked,
    isValidDate,
    // Context API
    FormConsumer,


    // Form Components
    CheckBoxField,
    EmailField,
    FileField,
    Form,
    PasswordField,
    RadioField,
    RadioGroup,
    SubmitButton,
    TextAreaField,
    TextInputField,
    SelectField,
    DatePickerField,
    QueryInputField,
};
export type {
    // Interfaces
    ICheckBoxField,
    IEmailField,
    IField,
    IPasswordField,
    IRadioField,
    ITextAreaField,
    ITextInputField,
    ISelectField,
    IDatePicker,
    IQueryField,
};

