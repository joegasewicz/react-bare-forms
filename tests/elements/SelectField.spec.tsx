import * as React from "react";
import {create} from "react-test-renderer";
import {Form, SelectField} from "../../src";
import { MockComponent } from "../form.spec";


describe("#<SelectField>", () => {
    it("should render props", () => {
        let state = {
            select_data_id: undefined as any,
        };

        let component = create(
            <MockComponent state={state}>
                <SelectField
                    size="lg"
                    value={state.select_data_id}
                    name="fruitChoice"
                    options={["banana", "apple", "orange"]}
                />
            </MockComponent>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("should render props from object keys", () => {
        let selectData: Array<{id: number, name: string}> = [
            {id: 1, name: "first"},
            {id: 2, name: "second"},
            {id: 3, name: "third"},
        ];
        let state = {
            select_data_id: undefined as any,
        };

        let component = create(
            <MockComponent state={state}>
                <SelectField
                    size="lg"
                    value={state.select_data_id}
                    name="fruitChoice"
                    objectkey="id"
                    objectvalue="name"
                    options={selectData}
                />
            </MockComponent>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
