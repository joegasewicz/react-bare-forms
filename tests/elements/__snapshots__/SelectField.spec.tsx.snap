// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#<SelectField> should render props 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <select
      className="form-control form-control-lg "
      name="fruitChoice"
      onChange={[Function]}
      options={
        Array [
          "banana",
          "apple",
          "orange",
        ]
      }
      size="lg"
    >
      <option
        value="banana"
      >
        banana
      </option>
      <option
        value="apple"
      >
        apple
      </option>
      <option
        value="orange"
      >
        orange
      </option>
    </select>
  </div>
</form>
`;

exports[`#<SelectField> should render props from object keys 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <select
      className="form-control form-control-lg "
      name="fruitChoice"
      objectKey="id"
      objectValue="name"
      onChange={[Function]}
      options={
        Array [
          Object {
            "id": 1,
            "name": "first",
          },
          Object {
            "id": 2,
            "name": "second",
          },
          Object {
            "id": 3,
            "name": "third",
          },
        ]
      }
      size="lg"
    >
      <option
        value={1}
      >
        first
      </option>
      <option
        value={2}
      >
        second
      </option>
      <option
        value={3}
      >
        third
      </option>
    </select>
  </div>
</form>
`;

exports[`<Form /> should render a form element with bootstrap extra tags 1`] = `
<form
  bare={false}
  context={[Function]}
  onSubmit={[Function]}
  state={
    Object {
      "username": "joebloggs",
    }
  }
/>
`;

exports[`<Form /> should render a form element with no extra tags 1`] = `
<form
  onSubmit={[Function]}
/>
`;
